tests/unit/package_managers/bundler/test_main.py

```diff
@@ -1,10 +1,92 @@
+    GemDependency,
+    GitDependency,
+    ParseResult,
+    PathDependency,
+)
 from cachi2.core.rooted_path import RootedPath
+from tests.common_utils import GIT_REF
+
+
+@mock.patch("cachi2.core.package_managers.bundler.main._get_main_package_name_and_version")
+@mock.patch("cachi2.core.package_managers.bundler.main.parse_lockfile")
+@mock.patch("cachi2.core.package_managers.bundler.parser.GemDependency.download_to")
+@mock.patch("cachi2.core.package_managers.bundler.parser.GitDependency.download_to")
+@mock.patch("cachi2.core.package_managers.bundler.parser.PathDependency.download_to")
+def test_resolve_bundler_package(
+    mock_path_dep_download_to: mock.Mock,
+    mock_git_dep_download_to: mock.Mock,
+    mock_gem_dep_download_to: mock.Mock,
+    mock_parse_lockfile: mock.Mock,
+    mock_get_main_package_name_and_version: mock.Mock,
+    rooted_tmp_path_repo: RootedPath,
+) -> None:
+    repo = Repo(rooted_tmp_path_repo)
+    repo.create_remote("origin", "git@github.com:user/repo.git")
+    repo_commit = repo.head.commit
+
+    package_dir = rooted_tmp_path_repo
+    output_dir = rooted_tmp_path_repo.join_within_root("cachi2-output")
+    deps_dir = output_dir.join_within_root("deps", "bundler")
+
+    gem_dep = GemDependency(
+        name="my-gem-dep",
+        version="0.1.0",
+        source="https://rubygems.org",
+    )
+    git_dep = GitDependency(
+        name="my-git-dep",
+        version="0.1.0",
+        url="https://github.com/rubygems/example.git",
+        ref=GIT_REF,
+    )
+    path_dep = PathDependency(
+        name="my-path-dep",
+        version="0.1.0",
+        root=package_dir,
+        subpath="vendor",
+    )
+
+    deps = [gem_dep, git_dep, path_dep]
+
+    mock_parse_lockfile.return_value = deps
+    mock_get_main_package_name_and_version.return_value = ("name", None)
+
+    components = _resolve_bundler_package(package_dir=package_dir, output_dir=output_dir)
+
+    mock_parse_lockfile.assert_called_once_with(package_dir)
+    mock_get_main_package_name_and_version.assert_called_once_with(package_dir, deps)
+    mock_gem_dep_download_to.assert_called_with(deps_dir)
+    mock_git_dep_download_to.assert_called_with(deps_dir)
+    mock_path_dep_download_to.assert_called_with(deps_dir)
+
+    assert len(components) == len(deps) + 1  # + 1 for the "main" package
+    assert deps_dir.path.exists()
+
+    components.sort(key=lambda c: c.name)
+    assert components[0].purl == "pkg:gem/my-gem-dep@0.1.0"
+    assert (
+        components[1].purl
+        == f"pkg:gem/my-git-dep@0.1.0?vcs_url=git%2Bhttps://github.com/rubygems/example.git%40{GIT_REF}"
+    )
+    assert (
+        components[2].purl
+        == f"pkg:gem/my-path-dep@0.1.0?vcs_url=git%2Bssh://git%40github.com/user/repo.git%40{repo_commit.hexsha}#vendor"
+    )
+    assert (
+        components[3].purl
+        == f"pkg:gem/name?vcs_url=git%2Bssh://git%40github.com/user/repo.git%40{repo_commit.hexsha}"
+    )
```

Comment:
This is unnecessarily difficult to read and understand. You can define a dedicated test case just for generating components with each type of component being an element of a parametrize. That way, you can directly stuff the expected PURL in the expected results and compare against that. It's going to be much more readable than this and not many lines will be needed for that (most lines will be "wasted" on component declaration in parametrize. Like I mentioned in my previous review, `test_generate_sbom_components` and `test_package_to_component` (rpm, gomod) can serve as an example here.